code/game/machinery/hologram.dm

```diff
@@ -41,313 +37,274 @@ var/const/HOLOPAD_MODE = RANGE_BASED
-		hologram.no_clean = TRUE
-		hologram.mouse_opacity = 0//So you can't click on it.
-		hologram.anchored = 1//So space wind cannot drag it.
-		masters[A] = hologram
-		A.holo = src
-		return 1
-
-/obj/machinery/hologram/holopad/proc/clear_holo(mob/living/silicon/ai/user, mob/living/carbon/call_user)
-	if(user)
-		qdel(masters[user])//Get rid of user's hologram
-		user.holo = null
-		masters -= user //Discard AI from the list of those who use holopad
-	if(call_user)
-		qdel(masters[call_user]) //Get rid of user's hologram
-		masters -= call_user //Discard the caller from the list of those who use holopad
-	if (!masters.len)//If no users left
-		set_light(0)			//pad lighting (hologram lighting will be handled automatically since its owner was deleted)
-		icon_state = "holopad0"
-		if(sourcepad)
-			sourcepad.targetpad = null
-			sourcepad = null
-			caller_id = null
-	return 1
 
-/obj/machinery/hologram/holopad/machinery_process()
-	for (var/mob/living/silicon/ai/master in masters)
-		var/active_ai = (master && !master.incapacitated() && master.client && master.eyeobj)//If there is an AI with an eye attached, it's not incapacitated, and it has a client
-		if((stat & NOPOWER) || !active_ai)
-			clear_holo(master)
+/obj/machinery/hologram/holopad/proc/create_holos()
+	for(var/mob/living/M in viewers(world.view, connected_pad))
+		if(LAZYISIN(active_holograms, M))
+			continue
+		create_holo(M)
+
+/obj/machinery/hologram/holopad/proc/create_holo(mob/M)
+	var/obj/effect/overlay/hologram/H = new(get_turf(src))
+	if(!isAI(M) && connected_pad)
+		H.x = src.x - (connected_pad.x - M.x)
+		H.y = src.y - (connected_pad.y - M.y)
+	if(!isInSight(H, src))
+		qdel(H)
+		return
+	H.assume_form(M)
+	LAZYSET(active_holograms, M, H)
+
+	update_icon()
+
+/obj/machinery/hologram/holopad/proc/update_holos()
+	for(var/thing in active_holograms)
+		var/mob/M = thing
+		if(isAI(M))
 			continue
+		var/obj/effect/overlay/hologram/H = active_holograms[thing]
+		if(connected_pad)
+			H.x = src.x - (connected_pad.x - M.x)
+			H.y = src.y - (connected_pad.y - M.y)
+		if(get_dist(H, src) > world.view)
+			clear_holo(M)
+			return
+		H.assume_form(M)
 
-		if(!(masters[master] in view(src)))
-			clear_holo(master)
+/obj/machinery/hologram/holopad/proc/clear_holos(var/clear_ai = TRUE)
+	for(var/M in active_holograms)
+		if(!clear_ai && isAI(M))
 			continue
+		clear_holo(M)
 
-		use_power(power_per_hologram)
-	if(last_request + 200 < world.time && incoming_connection==1)
-		incoming_connection = 0
-		end_call()
-		if(sourcepad)
-			sourcepad.audible_message("<i><span class='game say'>The holopad connection timed out</span></i>")
-			sourcepad = 0
-	if (caller_id && sourcepad)
-		if(caller_id.loc != sourcepad.loc)
-			forced = 0
-			sourcepad.visible_message("Severing connection to distant holopad.")
-			visible_message("The connection has been terminated by [caller_id].")
+/obj/machinery/hologram/holopad/proc/clear_holo(var/mob/M)
+	if(!length(active_holograms))
```

Comment:
You're using LAZY stuff elsewhere for this list, maybe use LAZYLEN here.

Reply:
length() works the same as LAZYLEN now, we can probably do a PR to get rid of it eventually

Reply:
I know but that's not the point of it. It is to indicate that the list is sometimes null. You use LAZY macros elsewhere and should here too.